lib/event_emitter/emitters/purchaser_order_stack_event_emitter.rb

```diff
@@ -0,0 +1,163 @@
+    end
+
+
+    module MessageBuilder
+      extend self
+
+      def build(purchaser_order:, purchaser_line_item:, purchaser:, operation:, backtrace:)
+        {
+          meta: meta(operation: operation, backtrace: backtrace),
+          payload: payload(
+            purchaser_order: purchaser_order,
+            purchaser_line_item: purchaser_line_item,
+            purchaser: purchaser,
+          ),
+        }
+      end
+
+      private
+
+      def meta(operation:, backtrace:)
+        {
+          message_uuid: UUIDTools::UUID.timestamp_create.to_s,
+          message_created_at: Time.now.utc.to_s,
+          message_operation: operation,
+          message_backtrace: backtrace,
+        }
+      end
+
+      def payload(purchaser_order:, purchaser_line_item:, purchaser:)
+        payload_section = {}
+        if purchaser_order
+          payload_section.merge! purchaser_order_details(
+                                   purchaser_order: purchaser_order,
+)
+        end
+        if purchaser_line_item
+          payload_section.merge! purchaser_line_item_details(
+                                   purchaser_line_item: purchaser_line_item,
+)
+        end
+        payload_section.merge! purchaser_details(purchaser: purchaser) if purchaser
+        payload_section
+      end
+
+      def purchaser_order_details(purchaser_order:)
+        {
+          purchaser_order: {
+            uuid: "3044b6f8-15b3-4b0d-98ef-7490c90cf9d3",
+            created_at: purchaser_order.created_at.utc.to_s,
+            order_reference: purchaser_order.retailers_order_ref,
+            purchaser_reference: purchaser_order.retailers_customer_ref,
+            organisation: {
+              uuid: "3347aace-172a-46b8-a4d9-d4f916628d1e",
+            },
+            purchaser_feed: {
+              uuid: "ac1875e1-25d9-46eb-83a0-e157b8db0356",
+            },
+          },
+        }
+      end
+
+      def purchaser_line_item_details(purchaser_line_item:)
+        {
+          purchaser_line_item: {
+            uuid: "92973dd1-d001-40ac-aa57-13eb2c963beb",
+            created_at: purchaser_line_item.created_at.utc.to_s,
+            product: {
+              uuid: "2dc110a9-c36d-4803-bc47-62fe4580b527",
+            },
+            purchaser_order: {
+              uuid: "3044b6f8-15b3-4b0d-98ef-7490c90cf9d3",
+            },
+            metadata: purchaser_line_item.metadata.to_json,
+          },
+        }
+      end
+
+      def purchaser_details(purchaser:)
+        {
+          purchaser: {
+            uuid: "ac1875e1-25d9-46eb-83a0-e157b8db0350",
+            is_unsubscribed: purchaser.unsubscribed?,
+            country_name: purchaser.country,
+            partial_postcode: purchaser.partial_postcode,
+            created_at: purchaser.created_at.utc.to_s,
+            updated_at: purchaser.created_at.utc.to_s,
+          },
+        }
+      end
+
+    end
+  end
+end
```

Comment:
This class that you created doesn't belong to the gem. The gem shouldn't know anything about what's the shape of payload. Shape of that payload should be specified somewhere else.

Now some of my thoughts (not sure if all will have sense)...

We could have a method in each model called `as_event`. You could call the method on a model and in return you would get a hash with all event data you need. Going further, maybe we could have a different gem which would store logic needed to create events, maybe even use something like json builder and have json event view for a model.

Imagine that to emit an event for creating a Purchaser record you call this:

```ruby
purchaser = Purchaser.find(123)
EventEmitter.new(:kinesis).publish(payload: purchaser.as_event(:create), options: options)
```

It requires creating an interface (a gem presumably), but I think it's worth it. 
Tell me your thoughts please!

Reply:
well originally was outside this gem (in revieworld), but John and Alex mentioned it would nice to have all the related code in the event_emitter gem rather than spread out in different repositories